### ISSUE TITLE
IPv6 reassembly + new packet definitions failed (solved: added a new packet clone method)

### ISSUE THREAD

**Issue by kbara** [Igalia/snabbswitch#393]
tl;dr: IPv6 reassembly worked on the private development branch, but when rebased onto the public branch exactly one test case mysteriously failed. This was due to a way of copying packets that doesn't work with our newly enriched packet structure. Fixed in https://github.com/snabbco/snabb/commit/edf6ef3224becd0fc1a47a45507932d180c60047

---

The mystery of the failing IPv4 fragmentation test case

I wrote a ctables-based hardened IPv6 reassembly implementation, and our end-to-end test suite passed. I rebased it onto our current dev branch, where the test suite also passed. The result failed exactly one test, which did an IPv6 reassembly followed by an IPv4 fragmentation. I hadn't changed anything that could touch the IPv4 fragmentation code. The problem wasn't the reassembly code: I added a new test which did only the reassembly, outputting a single IPv4 packet (using a larger MTU), and that passes on all the branches.

The one good thing about the test failure is that it is 100% reproducible. I set about analyzing and bisecting it. The test case produces three fragments; the first one lacks ethernet and IPv4 headers, and the third one contains two corrupt bytes; everything else is ok. There are several other IPv4 fragmentation tests, so it didn't look like a problem in the IPv4 fragmentation code.

Digging deeper into it, the first output packet is ok until a particular line in the source code:

```
   ffi.copy(last_pkt.data + header_size,
            ipv4_pkt.data + header_size + raw_frag_offset,
            last_payload_len)
```

This line seems fine: last_pkt and ipv4_packet are distinct packets (last_packet is created with packet.allocate(), and the two packets have memory addresses nowhere near each other), and header_size, raw_frag_offset, and last_payload_len all look ok. `ffi.copy` takes its destination argument first. And yet, the _source_ packet is corrupt after this line.

Packets have a pointer to a data buffer. I printed out the addresses of these data buffers, and got an interesting surprise:

```
cdata<unsigned char *>: 0x5000039e37f8
cdata<unsigned char *>: 0x5000039e37d0
```

The data buffers start only 40 bytes apart! This is problematic, as they're about 10,000 bytes long... and it explains how a copy is corrupting the source buffer.

The question arises as to how the buffers of two different packets can overlap. In upstream Snabb, they can't; packets have inlined data, and are defined as

```
struct packet {
    unsigned char data[PACKET_PAYLOAD_SIZE];
    uint16_t length;           // data payload length
};
```

Restoring upstream Snabb's packet.h and packet.lua makes the problem vanish, and the test suite pass. But that's not a satisfactory answer; our development branch has a more complicated structure for the sake of major performance gains with virtualization:

```
struct packet {
    unsigned char *data;
    uint16_t length;           // data payload length
    uint16_t headroom;         // payload starts this many bytes into data_
    uint32_t padding;
    unsigned char data_[PACKET_PAYLOAD_SIZE];
};
```

This clearly comes with the intent that data be a pointer into data_. And yet, printing the address of the first packet and its data buffer tells a sinister tale:....

```
The packet: cdata<struct packet *>: 0x7fda4a555d3c 
The data: cdata<unsigned char *>: 0x5000039e37f8 
```

At this point, it's clear that the fragmentation test failure was just a very lucky symptom of a much deeper bug. That bug turns out to be a combination of several design decisions:
- The extra level of indirection in packet.lua
- A line inserted into the IPv6 reassembly code, to inline the reassembly packet into the underlying data structure.
- The use of ffi.copy for the inlining:

```
ffi.copy(reassembly_buf.reassembly_packet, pkt, struct_packet_size)
```

With the new definition of a packet, that guarantees that `data` will point to a buffer in the wrong packet. This is compounded by freeing the now-copied packet right away. The next step of the test, which does IPv4 fragmentation, reuses the packet (via a pool mechanism baked into Snabb), and suddenly, the error becomes visible.

The solution is a method to clone a packet into pre-allocated packet structure, which carefully does all the paperwork suitable to the current definition of a packet, in packet.lua, rather than using ffi.copy.

The main lesson is that end to end test suites are extremely useful. All of the individual units were working as expected, but as the implementation details of one unit (the packet data structure) changed, assumptions in code that used it broke, in a way that was invisible to unit tests of the IPv6 code. Assumptions between modules have always been one of the richest sources of bugs; testing those assumptions is one very valuable.
